### DISCUSSION TITLE
Decision Tables

### DISCUSSION THREAD

**Discussion by josesuero** [serverlessworkflow/specification#741] [Ideas]
I really love this initiative, and was able to get buy in from my team mainly because our alternative was BPMN and having a JSON understandable format was a clear win.

Is there any initiatives to do the same with DMN? currently we have several decision tables and would love to have the same approach as workflows for decisions. Change clunky XML with understandable JSON.

Maybe another project is tackling this, if so, please let me know which.

What do you think?

**Comment by cdavernas** [comment]
Hello @josesuero!

As far as I know, depending on the complexity of your DMN tables, you could achieve the same using the ServerlessWorkflow's data-based [`switch` state](https://github.com/serverlessworkflow/specification/blob/main/specification.md#switch-state) on top of `jq` or any other runtime expression language.

**Comment by josesuero** [comment]
Hey @cdavernas 

For basic decisions switch work, same as in BPMN you can add routes based on comparing fields. But the decisions I'm talking about are the ones that compare multiple fields with multiple conditions and multiple levels. Like modeling German tax law or stock market decisions.

DMN work really well to model these kind of projects, because they're super easy to make, and also great to show stakeholders. 

This is one the things I adore about serverless workflows, I can easily make them, my teammates can easily check changes on pull requests, and the SVG makes sense to stake holders. Something like this to replace DMNs would be interesting to consider.

**Reply by ricardozanini** [reply]
Why not externalize these DMNs tables in other services and orchestrate them via Serverless Workflow?

**Reply by tsurdilo** [reply]
Think idea is to express the decision logic as json/yaml and have open source schema for it. DSL is imo well suited for this. Adding event-based logic to decisions is imo also very interesting and useful maybe. Could build out workflow logic based on this input too. Very interesting.

**Reply by tarilabs** [reply]
Would something like this: https://github.com/kiegroup/yard#simple-example
be of interest?

To be noted:

- yaml with semantic concepts of Decision Table and expression, per original post.
- supports `jq` (https://stedolan.github.io/jq) for expressing constraints which I understand is default expression language for CNCF's SW: [example](https://github.com/kiegroup/yard#plug-in-mechanism-for-multiple-expression-languages)
- we have explored integration into cloud-native for instance as CRD for K8s
- that is not tied to DMN, but borrows some interesting aspect from it

Hope this is of interest? :)

**Reply by tsurdilo** [reply]
Hi @tarilabs do you mean as to being able to process the sample yaml you shared as workflow input? I think SW would allow you to do that already. Trying to understand how you see this integration?

**Reply by tarilabs** [reply]
Yes. I meant the capability I pointed to ("yard") could be used from/within a SW to invoke a decision table evaluation expressed as yaml, and the result of that could be used to decide which branch of the workflow you want to follow next in the SW definition. Remotely as a service on a k8s cluster or "locally" to the SW I believe SW provides technical support for both scenario already for those type of integrations.

**Reply by tsurdilo** [reply]
got it, ok so use it as a service.

yeah i mean with SW you can definitely integrate with other services similar to what you describe. thanks for sharing

**Reply by tarilabs** [reply]
> got it, ok so use it as a service.
> yeah i mean with SW you can definitely integrate with other services similar to what you describe. thanks for sharing

thank you for the feedback; if you see other type of integration/service in this context don't hesitate to share I'd be glad to learn
